[CASSANDRA-18302] Statement source descriptions

jacek-lewandowski opened the pull request:
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

dcapwell commented on src/java/org/apache/cassandra/cql3/StatementSource.java:
```diff
@@ -0,0 +1,162 @@
+        private boolean isEmpty()
+        {
+            return startLine == Character.MAX_VALUE ||
+                   startLine > endLine ||
```
what is the case that this happens?  this feels like a bug, should we not fail?

jacek-lewandowski replied:
In all those pointed out cases I tried to make this class super resistant to wrong input regardless of the circumstances. But it can fail as well.

dcapwell commented on src/java/org/apache/cassandra/cql3/StatementSource.java:
```diff
@@ -0,0 +1,162 @@
+        {
+            return startLine == Character.MAX_VALUE ||
+                   startLine > endLine ||
+                   startLine == endLine && (startChar == Character.MAX_VALUE || startChar > endChar);
```
`startChar > endChar` feels like a bug

dcapwell commented on src/java/org/apache/cassandra/cql3/StatementSource.java:
```diff
@@ -0,0 +1,162 @@
+
+        if (startToken.getType() == Token.EOF)
+        {
+            startLine = Character.MAX_VALUE;
```
if start token is EOF do we not know that endToken is as well?

jacek-lewandowski replied:
Yes, we can assume that

dcapwell commented on src/java/org/apache/cassandra/cql3/statements/DeleteStatement.java:
```diff
@@ -174,7 +186,8 @@ protected ModificationStatement prepareInternal(ClientState state,
                                                        conditions,
-                                                       attrs);
+                                                       attrs,
+                                                       bounds.describeSource(enclosingStatement));
```
this feels backwards, we create the underline statements before we create the composite statement, so how do we pass it in?

jacek-lewandowski replied:
No, notice that the enclosing composite statement is a parsed statement and it already exists so as the parsed delete statement.

dcapwell commented on src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java:
```diff
@@ -207,12 +204,9 @@ public ResultSet.ResultMetadata getResultMetadata()
         // We reject reads from both LET and SELECT that do not specify a single row.
         @SuppressWarnings("unchecked")
-        SinglePartitionReadQuery.Group<SinglePartitionReadCommand> selectQuery = (SinglePartitionReadQuery.Group<SinglePartitionReadCommand>) readQuery;
+        SinglePartitionReadQuery.Group<SinglePartitionReadCommand> selectQuery = (SinglePartitionReadQuery.Group<SinglePartitionReadCommand>) select.getQuery(options, 0);
```
why prefer ClassCast rather than `ILLEGAL_RANGE_QUERY_MESSAGE`?

jacek-lewandowski replied:
In this and the other case, the statement at this point is guaranteed to not be partition range query. It has been verified during preparation.

dcapwell commented on src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java:
```diff
@@ -223,12 +217,9 @@ TxnNamedRead createNamedRead(NamedSelect namedSelect, QueryOptions options, Clie
         // We reject reads from both LET and SELECT that do not specify a single row.
         @SuppressWarnings("unchecked")
-        SinglePartitionReadQuery.Group<SinglePartitionReadCommand> selectQuery = (SinglePartitionReadQuery.Group<SinglePartitionReadCommand>) readQuery;
+        SinglePartitionReadQuery.Group<SinglePartitionReadCommand> selectQuery = (SinglePartitionReadQuery.Group<SinglePartitionReadCommand>) select.getQuery(options, 0);
```
why prefer class cast over `ILLEGAL_RANGE_QUERY_MESSAGE`?

dcapwell reviewed (commented):
got to this late and there is parts that I feel I need to poke with an IDE to properly review, so I sadly have to delay real review to tomorrow...

jacek-lewandowski commented:
thanks @dcapwell, I've applied some fixes basing on your feedback

maedhroz commented on src/java/org/apache/cassandra/cql3/statements/QualifiedStatement.java:
```diff
@@ -81,13 +81,27 @@ public String toString()
 
+        public void setRawText(String txt)
+        {
+            this.txt = txt != null ? txt.substring(0, Math.min(txt.length(), RAW_TEXT_MAX_LENGTH)) : null;
```
nit: if we actually do exceed the max length, it might be a nice UX tweak to include a "blah blah..." (ellipsis) at the end just to make it clear there's more content (but we're truncating).

maedhroz commented on src/java/org/apache/cassandra/cql3/statements/BatchStatement.java:
```diff
@@ -235,7 +261,7 @@ public void validate() throws InvalidRequestException
             {
                 if (ksName != null && (!stmt.keyspace().equals(ksName) || !stmt.table().equals(cfName)))
-                    throw new InvalidRequestException("Batch with conditions cannot span multiple tables");
+                    throw new InvalidRequestException("Batch with conditions cannot span multiple tables: " + stmt.source.describe(true));
```
Is there any coverage for these two cases in `BatchTest`? In general, feels like it would be good to have a test for the `ModificationStatement` case...

maedhroz commented on src/java/org/apache/cassandra/cql3/statements/QualifiedStatement.java:
```diff
@@ -81,13 +81,27 @@ public String toString()
 
     public static abstract class Composite extends QualifiedStatement
     {
+        private static final int RAW_TEXT_MAX_LENGTH = 4096;
```
If we set this to something lower (let's say 38), and then run `shouldRejectIllegalLimitInLet()`, you encounter a situation where the `Bounds` points to non-existent raw data. I suppose this won't be problematic unless we have very large statements.

In `CQLFragmentParser#parseAnyUnhandled()`, I think we already have the `Bounds` objects we would later use to filter the raw text, right? What if, instead of filtering later, and storing the raw text for the entire enclosing statement, we filter each statement in `CQLFragmentParser#parseAnyUnhandled()` avoiding the need for a limit in the first place? The individual statement would just get the simple bounds and the text from the raw statement text within those bounds...and we'd be able to present the error the same way (I think).

WDYT?

maedhroz commented on src/java/org/apache/cassandra/cql3/StatementSource.java:
```diff
@@ -0,0 +1,162 @@
+        else
+        {
+            if (!bounds.isEmpty() && text != null && mayIncludeSourceText)
+                return String.format("at %d:%d - %s", bounds.startLine + 1, bounds.startChar + 1, text);
```
nit: Another way to do it that could avoid the dash...

`at [2:1] SELECT....`

Matter of style though, so feel free to ignore.

dcapwell commented on src/java/org/apache/cassandra/cql3/StatementSource.java:
```diff
@@ -0,0 +1,87 @@
+        int startLine = min(max(startToken.getLine(), 1) - 1, Character.MAX_VALUE);
+        int startChar = min(max(startToken.getCharPositionInLine(), 0), Character.MAX_VALUE);
+        String txt = input.toString(startToken, endToken);
+        txt = txt.replace('\n', ' ');
```
we should keep the newline,

```
SELECT a,
               b,
               c -- should fail at line 3 due to 3 not being in the schema
```

dcapwell commented on src/java/org/apache/cassandra/cql3/StatementSource.java:
```diff
@@ -0,0 +1,87 @@
+        int startChar = min(max(startToken.getCharPositionInLine(), 0), Character.MAX_VALUE);
+        String txt = input.toString(startToken, endToken);
+        txt = txt.replace('\n', ' ');
+        txt = TWO_OR_MORE_WHITESPACES.matcher(txt).replaceAll(" ").trim();
```
we should leave the input alone, if we want to show the string the users provided, we should show as-is

maedhroz commented on src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java:
```diff
@@ -366,21 +357,19 @@ public ResultMessage execute(QueryState state, QueryOptions options, long queryS
 
             if (returningSelect != null)
-                checkAtMostOneRowSpecified(state.getClientState(), options, returningSelect.select, INCOMPLETE_PRIMARY_KEY_SELECT_MESSAGE);
+                checkFalse(isSelectingMultipleClusterings(returningSelect.select, options), INCOMPLETE_PRIMARY_KEY_SELECT_MESSAGE, "returning SELECT", returningSelect.select.source);
```
I'd even be fine if we just did what we do for INSERT/UPDATE...that is, just say `SELECT at [2:1]` or `LET at [1:1]`. Not sure if someone reading this would glean information from "returning" or it's just an implementation detail.

CC @dcapwell

dcapwell replied:
"returning select" I think helps as we can only have 1 of those

maedhroz replied:
The only other SELECTs would be called LET here...anyway...I'm good with whatever the two of you like better.

maedhroz commented on test/unit/org/apache/cassandra/cql3/StatementSourceTest.java:
```diff
@@ -0,0 +1,65 @@
+
+public class StatementSourceTest
+{
+    private final static Logger logger = LoggerFactory.getLogger(StatementSourceTest.class);
```
logger unused now

maedhroz reviewed (approved):
LGTM

jacek-lewandowski committed:
Improve transaction statement validation

patch by Jacek Lewandowski; reviewed by David Capwell and Caleb Rackliffe for CASSANDRA-18302
